### PR TITLE
fix: handle InvalidLabelChar in URL validation to prevent panic

### LINKED ISSUE #23294: thread 'main' panicked at runtime/permissions/lib.rs:695:19
Platform: linux x86_64
Version: 1.42.1
Args: ["deno", "run", "icallersa.js"]

thread 'main' panicked at runtime/permissions/lib.rs:695:19:
called `Result::unwrap()` on an `Err` value: InvalidLabelChar
stack backtrace:
   0: rust_begin_unwind
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/panicking.rs:72:14
   2: core::result::unwrap_failed
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/result.rs:1649:5
   3: core::result::Result<T,E>::unwrap
   4: deno_permissions::UnaryPermission<deno_permissions::NetDescriptor>::check_url
   5: deno_fetch::op_fetch<FP>::v8_fn_ptr
   6: Builtins_CallApiCallbackGeneric

### LINKED ISSUE #23552: Panic when passing wildcard to allow-net
Platform: linux x86_64
Version: 1.42.1
Args: ["deno", "run", "--allow-read", "--allow-write=notion-next", "--allow-net=api.notion.com,*.amazonaws.com", "--allow-env", "main.ts"]

thread 'main' panicked at runtime/permissions/lib.rs:741:22:
called `Result::unwrap()` on an `Err` value: InvalidLabelChar

### PR DESCRIPTION
**Pull request by yazan-abdalrahman**
Fixes #23294
Fixes #23552

### CODE REVIEW

**Comment by yazan-abdalrahman** [comment]
before
image

after
image

File: ext/net/ops.rs [hunk 1]
```diff
@@ -59,6 +63,40 @@ pub struct IpAddr {
...
+/// # Returns
+///
+/// * `Result<(), AnyError>` - Returns `Ok(())` if the hostname and port are successfully
+///   validated and updated, otherwise returns an error.
+///
+/// # Errors
+///
+/// This function will return an error if the hostname cannot be parsed or if any validation fails.
+///
+fn process_ip_addr(addr: &mut IpAddr) -> Result<(), AnyError> {
```
Comment (dsherret) [inline ext/net/ops.rs:86]: `process_ip_addr` -> `normalize_ip_addr`?

File: ext/net/ops.rs [hunk 2]
```diff
@@ -59,6 +63,40 @@ pub struct IpAddr {
   pub port: u16,
 }
 
+/// Processes and validates the hostname and port in the given `IpAddr` structure.
+///
+/// This function extracts the hostname and port from the provided `IpAddr` structure,
+/// validates them, and updates the structure with the correctly formatted hostname and port.
+/// This is necessary to ensure that the hostname and port are valid before they are used
+/// in network operations such as `Deno.connect` and `Deno.listen`.
```
Comment (dsherret) [inline ext/net/ops.rs:71]: Do we need this function for this PR? It seems like a new feature, which I'm not sure we should do. Can we land this change in a separate PR?

Comment (yazan-abdalrahman) [reply ext/net/ops.rs:71]: Yes, it is a new addition, however it is not complex or can be described as a new feature; it is simply an upgrade of existing behavior. 

**This is to extract the hostname from a URL and allow permission to access the hostname at a certain port, not the entire URL. For example, https://url:port/.


ity will offer permission for url at port from url rather than giving port at connect method. 


If we don't put:port, it will use the port argument in deno connect or deno listen.**

File: ext/net/ops.rs (hunk 1 above)
Comment (yazan-abdalrahman) [reply ext/net/ops.rs:86]: OK

**Comment by yazan-abdalrahman** [comment]
@dsherret 

After updating the branch and resolving the merge conflict issue, I encountered a cyclic dependence issue; how can we resolve it? I need to change  NetPermissions  place to avoid cycle dependence issue?

image

**Comment by yazan-abdalrahman** [comment]
@dsherret 

After updating the branch and resolving the merge conflict issue, I encountered a cyclic dependence issue; how can we resolve it? I need to change  NetPermissions  place to avoid cycle dependence issue?

image
image

error: cyclic package dependency: package `deno_permissions v0.15.0 (D:\yad\deno\runtime\permissions)` depends on itself. Cycle:
package `deno_permissions v0.15.0 (D:\yad\deno\runtime\permissions)`
    ... which satisfies path dependency `deno_permissions` (locked to 0.15.0) of package `deno_net v0.147.0 (D:\yad\deno\ext\net)`
    ... which satisfies path dependency `deno_net` (locked to 0.147.0) of package `deno_permissions v0.15.0 (D:\yad\deno\runtime\permissions)`
    ... which satisfies path dependency `deno_permissions` (locked to 0.15.0) of package `deno_web v0.186.0 (D:\yad\deno\ext\web)`
    ... which satisfies path dependency `deno_web` (locked to 0.186.0) of package `deno_crypto v0.169.0 (D:\yad\deno\ext\crypto)`
    ... which satisfies path dependency `deno_crypto` (locked to 0.169.0) of package `deno_runtime v0.163.0 (D:\yad\deno\runtime)`
    ... which satisfies path dependency `deno_runtime` (locked to 0.163.0) of package `deno v1.44.1 (D:\yad\deno\cli)`

**Comment by yazan-abdalrahman** [comment]
@dsherret 

After updating the branch and resolving the merge conflict issue, I encountered a cyclic dependence issue; how can we resolve it? I need to change  NetPermissions  place to avoid cycle dependence issue?

image
image

error: cyclic package dependency: package `deno_permissions v0.15.0 (D:\yad\deno\runtime\permissions)` depends on itself. Cycle:
package `deno_permissions v0.15.0 (D:\yad\deno\runtime\permissions)`
    ... which satisfies path dependency `deno_permissions` (locked to 0.15.0) of package `deno_net v0.147.0 (D:\yad\deno\ext\net)`
    ... which satisfies path dependency `deno_net` (locked to 0.147.0) of package `deno_permissions v0.15.0 (D:\yad\deno\runtime\permissions)`
    ... which satisfies path dependency `deno_permissions` (locked to 0.15.0) of package `deno_web v0.186.0 (D:\yad\deno\ext\web)`
    ... which satisfies path dependency `deno_web` (locked to 0.186.0) of package `deno_crypto v0.169.0 (D:\yad\deno\ext\crypto)`
    ... which satisfies path dependency `deno_crypto` (locked to 0.169.0) of package `deno_runtime v0.163.0 (D:\yad\deno\runtime)`
    ... which satisfies path dependency `deno_runtime` (locked to 0.163.0) of package `deno v1.44.1 (D:\yad\deno\cli)`

**Comment by yazan-abdalrahman** [comment]
@dsherret 

After updating the branch and resolving the merge conflict issue, I encountered a cyclic dependence issue; how can we resolve it? I need to change  NetPermissions  place to avoid cycle dependence issue?

image

File: runtime/permissions.rs
Comment (dsherret) [inline runtime/permissions.rs:1]: I think this file can be reverted? It was removed from main.

File: ext/net/ops.rs (hunk 2 above)
Comment (dsherret) [reply ext/net/ops.rs:71]: I'd rather we not upgrade the existing behaviour for this bug fix and instead land that in a separate PR.

File: runtime/permissions/lib.rs [hunk 3]
```diff
@@ -1142,17 +1155,14 @@ impl UnaryPermission<NetDescriptor> {
...
-      .ok_or_else(|| uri_error("Missing host"))?
-      .to_string();
-    let host = &(&hostname, url.port_or_known_default());
-    let display_host = match url.port() {
-      None => hostname.clone(),
-      Some(port) => format!("{hostname}:{port}"),
-    };
-    self.check_desc(Some(&NetDescriptor::new(&host)), false, api_name, || {
-      Some(format!("\"{}\"", display_host))
+    dbg!(url.clone());
```
Comment (dsherret) [inline runtime/permissions/lib.rs:1158]: I think this is causing some test failures.

File: runtime/permissions.rs
Comment (dsherret) [reply runtime/permissions.rs:1]: Seems still relevant?

File: ext/net/ops.rs (hunk 2 above)
Comment (dsherret) [reply ext/net/ops.rs:71]: I unresolved this comment—seems still relevant?

File: runtime/permissions/lib.rs [hunk 4]
```diff
@@ -3331,4 +3520,114 @@ mod tests {
...
+      NetDescriptor::from_cli_arg("[2606:4700:4700::1111]:80").unwrap(),
+      NetDescriptor(
+        Host::Ipv6(Ipv6Addr::new(0x2606, 0x4700, 0x4700, 0, 0, 0, 0, 0x1111)),
+        Some(80)
+      )
+    );
+
+    // Parsing a URL with a host
+    assert_eq!(
+      NetDescriptor::from_cli_arg("https://github.com/denoland/").unwrap(),
```
Comment (dsherret) [inline runtime/permissions/lib.rs:3567]: Can you revert this new functionality? We should try to make this PR as similar to what it was like before minus the panic.

Comment (yazan-abdalrahman) [reply runtime/permissions/lib.rs:3567]: OK

File: runtime/permissions.rs
Comment (yazan-abdalrahman) [reply runtime/permissions.rs:1]: I revert my change from this file, but file it still existed, therefore I reverted the entire file from my change in the last commit.

**Comment by yazan-abdalrahman** [comment]
> See comment about removing new functionality.

done I deleted parsing URL as the origin behavior, mapped error to URIError as the origin instance of error and removed the file you mentioned as not relevant.

**Comment by yazan-abdalrahman** [comment]
> See comment about removing new functionality.

@dsherret 
I removed the new feature of parsing URLs and the origin behavior in --allow-net=github.com:443. It gives permission for github.com at port 443 only, but when you do --allow-net=github.com, it will give permission at all ports, which is the original behavior, and this behavior works with IPv4, IPv6, and hostname.

I saw that via this reference:
https://docs.deno.com/runtime/manual/basics/permissions
https://deno.land/api@v1.44.4?s=Deno.connect
https://deno.land/api@v1.44.4?s=Deno.listen 


If you need to remove any functions, it will defect the permission system.

File: ext/net/ops.rs (hunk 2 above)
Comment (yazan-abdalrahman) [reply ext/net/ops.rs:70]: This function is for the permission system because after I made changes to parsing the host and validating it to avoid the panic in allow-net, the permission system was affected, therefore I fixed it by introducing this function. and this not new feature

see this reference:
https://docs.deno.com/runtime/manual/basics/permissions
https://deno.land/api@v1.44.4?s=Deno.connect
https://deno.land/api@v1.44.4?s=Deno.listen

**Comment by yazan-abdalrahman** [comment]
@dsherret 
Hello, can you re-run the pipeline

**Comment by bartlomieju** [comment]
Closing in favor of https://github.com/denoland/deno/pull/24397 which landed parts of this PR.
